lib/srv/db/mysql/gcp.go

```diff
@@ -0,0 +1,138 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package mysql
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/cloud/gcp"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+)
+
+func isDBUserGCPServiceAccount(dbUser string) bool {
+	if strings.Contains(dbUser, "@") {
+		switch {
+		// Example: mysql-iam-user@my-project-id.iam
+		// This format is used to align with PostgreSQL.
+		case strings.HasSuffix(dbUser, ".iam"):
+			return true
+		// Example: mysql-iam-user@my-project-id..gserviceaccount.com
+		case strings.HasSuffix(dbUser, ".iam.gserviceaccount.com"):
+			return true
+		}
+	}
+	return false
+}
+
+func gcpServiceAccountToDatabaseUser(serviceAccountName string) string {
+	user, _, _ := strings.Cut(serviceAccountName, "@")
+	return user
+}
+
+func databaseUserToGCPServiceAccount(sessionCtx *common.Session) string {
+	return fmt.Sprintf("%s@%s.iam.gserviceaccount.com", sessionCtx.DatabaseUser, sessionCtx.Database.GetGCP().ProjectID)
+}
+
+func (e *Engine) getGCPUserAndPassword(ctx context.Context, sessionCtx *common.Session, gcpClient gcp.SQLAdminClient) (string, string, error) {
+	// If `--db-user` is an service account email, use IAM Auth.
+	if isDBUserGCPServiceAccount(sessionCtx.DatabaseUser) {
+		return e.getGCPIAMUserAndPassword(ctx, sessionCtx)
+	}
+
+	// Get user info to decide how to authenticate.
+	user, err := gcpClient.GetUser(ctx, sessionCtx.Database, sessionCtx.DatabaseUser)
+	if err != nil {
+		// GetUser permission is new for IAM auth. If no permission, assume legacy password user.
+		if trace.IsAccessDenied(err) {
+			return e.getGCPPasswordUserAndPassword(ctx, sessionCtx)
+		}
+		return "", "", trace.Wrap(err)
+	}
+
+	// Possible values (copied from SDK):
+	//   "BUILT_IN" - The database's built-in user type.
+	//   "CLOUD_IAM_USER" - Cloud IAM user.
+	//   "CLOUD_IAM_SERVICE_ACCOUNT" - Cloud IAM service account.
+	//   "CLOUD_IAM_GROUP" - Cloud IAM group non-login user.
+	//   "CLOUD_IAM_GROUP_USER" - Cloud IAM group login user.
+	//   "CLOUD_IAM_GROUP_SERVICE_ACCOUNT" - Cloud IAM group service account.
+	//
+	// In practice, type can also be empty for built-in user.
+	switch user.Type {
+	case "",
+		"BUILT_IN":
+		return e.getGCPPasswordUserAndPassword(ctx, sessionCtx)
+
+	case "CLOUD_IAM_SERVICE_ACCOUNT",
+		"CLOUD_IAM_GROUP_SERVICE_ACCOUNT":
+		return e.getGCPIAMUserAndPassword(
+			ctx,
+			sessionCtx.WithUser(databaseUserToGCPServiceAccount(sessionCtx)),
+		)
+
```

Comment:
Can you move enums to constans for readabilty

Reply:
added